presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java

```diff
@@ -47,31 +49,37 @@ public LocalExchangeMemoryManager(long maxBufferedBytes)
 
     public void updateMemoryUsage(long bytesAdded)
     {
-        SettableFuture<?> future;
-        synchronized (this) {
-            bufferedBytes.addAndGet(bytesAdded);
-
-            // if we are full, then breakout
-            if (bufferedBytes.get() > maxBufferedBytes || notFullFuture.isDone()) {
-                return;
+        long bufferedBytes = this.bufferedBytes.addAndGet(bytesAdded);
+        // detect the transition from above to below the full boundary
+        if (bufferedBytes <= maxBufferedBytes && (bufferedBytes - bytesAdded) > maxBufferedBytes) {
+            SettableFuture<?> future;
+            synchronized (this) {
+                //  if we have no callback waiting, return early
+                if (notFullFuture == NOT_FULL) {
+                    return;
+                }
+                future = notFullFuture;
+                notFullFuture = NOT_FULL;
             }
-
-            // otherwise, we are not full, so complete the future
-            future = notFullFuture;
-            notFullFuture = NOT_FULL;
+            // complete future outside of lock since this can invoke callbacks
+            future.set(null);
         }
-
-        // complete future outside of lock since this can invoke callbacks
-        future.set(null);
     }
 
-    public synchronized ListenableFuture<?> getNotFullFuture()
+    public ListenableFuture<?> getNotFullFuture()
     {
-        // if we are full and the current not full future is already complete, create a new one
-        if (bufferedBytes.get() > maxBufferedBytes && notFullFuture.isDone()) {
-            notFullFuture = SettableFuture.create();
+        if (bufferedBytes.get() <= maxBufferedBytes) {
+            return NOT_BLOCKED;
+        }
+        SettableFuture<?> future;
+        synchronized (this) {
+            // if we are full and the current not full future is already complete, create a new one
+            if ((notFullFuture == NOT_FULL || notFullFuture.isDone()) && bufferedBytes.get() > maxBufferedBytes) {
```

Comment:
simplify:
```
// optionally
if (bufferedBytes.get() <= maxBufferedBytes) {
  return NOT_BLOCKED;
}
if (notFullFuture.isDone()) {
  notFullFuture = SettableFuture.create();
}
```

Reply:
Simplified, also realized that the `notFullFuture.isDone()` check is not correct as it can race with the callback. Updated to now only use the  `notFullFuture == NOT_DONE` alias check for callback registration.

Reply:
Updated again to use a null sentinel and this becomes a null check.